File: packages/python/goldenpipe/goldenpipe/core/_planner_json.py
"""The JSON face of the pure-Python planner — the SP2 parity surface (SHIPPED).

Each ``*_json`` fn CALLS the real Resolver/Router/decisions/_auto_config and serializes
to goldenpipe-core's exact JSON shapes, so the parity gate tests the actual planner
(not a re-implementation). It does NOT run at pipeline runtime; ordering/validation/
routing stay in the engine modules. Mirrors goldenpipe-core/src/json.rs.
"""
from __future__ import annotations

import json
from typing import Any

from goldenpipe import decisions as _dec
from goldenpipe.engine.resolver import PlannedStage, Resolver, WiringError
from goldenpipe.engine.router import Router
from goldenpipe.models.config import PipelineConfig, StageSpec
from goldenpipe.models.context import Decision, PipeContext
from goldenpipe.models.stage import StageInfo


class _Stub:
    """A minimal Stage object carrying .info (all the planner reads)."""

    def __init__(self, info: StageInfo) -> None:
        self.info = info


class _StubRegistry:
    """Registry keyed EXPLICITLY (bypasses register()'s key-by-info.name) so
    ``key != info.name`` vectors resolve. Provides get()/list_all()."""

    def __init__(self) -> None:
        self._stages: dict[str, Any] = {}

    def add(self, key: str, info: StageInfo) -> None:
        self._stages[key] = _Stub(info)

    def get(self, name: str) -> Any:
        if name not in self._stages:
            raise KeyError(f"Stage '{name}' not found in registry")
        return self._stages[name]

    def list_all(self) -> dict[str, StageInfo]:
        return {k: s.info for k, s in self._stages.items()}


def _info(d: dict) -> StageInfo:
    return StageInfo(name=d["name"], produces=list(d["produces"]), consumes=list(d["consumes"]))


def _planned_to_dict(p: Any) -> dict:
    out = {"name": p.name, "use": p.spec.use, "config": p.config or {}, "on_error": p.spec.on_error}
    if p.spec.skip_if is not None:
        out["skip_if"] = p.spec.skip_if
    return out


def resolve_json(input_str: str) -> str:
    arg = json.loads(input_str)
    reg = _StubRegistry()
    for s in arg["stages"]:
        reg.add(s["key"], _info(s))  # key by the registry KEY, not info.name
    config = PipelineConfig(**arg["config"])
    try:
        plan = Resolver.resolve(config, reg)
    except WiringError as e:
        return json.dumps(
            {"err": {"kind": "wiring", "stage": e.stage, "missing": e.missing, "available": e.available}}
        )
    except KeyError:
        # unknown `use`: the first configured stage whose use isn't registered
        # (Resolver fails on the first such stage, in order).
        for raw in config.stages:
            use = raw if isinstance(raw, str) else raw.use
            if use not in reg._stages:
                return json.dumps({"err": {"kind": "unknown_stage", "use": use}})
        raise  # unreachable
    return json.dumps({"ok": {"stages": [_planned_to_dict(p) for p in plan.stages]}})


def apply_decision_json(input_str: str) -> str:
    arg = json.loads(input_str)
    d = arg["decision"]
    decision = Decision(
        skip=d.get("skip", []),
        abort=d.get("abort", False),
        insert=d.get("insert", []),
        reason=d.get("reason", ""),
    )
    remaining = []
    for r in arg["remaining"]:
        # carry on_error/skip_if through the spec so a remaining stage that has them
        # round-trips faithfully (the core's PlannedSpec round-trips both).
        spec = StageSpec(
            use=r["use"], name=r.get("name"), on_error=r.get("on_error", "continue"),
            skip_if=r.get("skip_if"),
        )
        remaining.append(PlannedStage(name=r["name"], stage=None, spec=spec, config=r.get("config", {})))
    reg = _StubRegistry()
    for name in decision.insert:  # Router.get(name) for each inserted stage
        reg.add(name, StageInfo(name=name, produces=[], consumes=[]))
    ctx = PipeContext()
    new_remaining = Router.apply(decision, remaining, ctx, reg)
    out: dict = {"remaining": [_planned_to_dict(p) for p in new_remaining]}
    note = ctx.reasoning.get("_router")
    if note is not None:
        out["router_note"] = note
    return json.dumps(out)


_BUILTINS = {
    "severity_gate": _dec.severity_gate,
    "pii_router": _dec.pii_router,
    "row_count_gate": _dec.row_count_gate,
}


def evaluate_builtin_json(input_str: str) -> str:
    arg = json.loads(input_str)
    fn = _BUILTINS.get(arg["name"])
    if fn is None:
        return "null"
    ctx_in = arg.get("ctx", {})
    ctx = PipeContext(artifacts=ctx_in.get("artifacts", {}), metadata=ctx_in.get("metadata", {}))
    d = fn(ctx)
    if d is None:
        return "null"
    return json.dumps({"skip": d.skip, "abort": d.abort, "insert": d.insert, "reason": d.reason})


def auto_config_json(input_str: str) -> str:
    from goldenpipe.pipeline import Pipeline

    arg = json.loads(input_str)
    reg = _StubRegistry()
    for name in arg["available"]:
        reg.add(name, StageInfo(name=name, produces=[], consumes=[]))
    p = Pipeline(registry=reg, identity_opts=arg.get("identity_opts"))
    cfg = p._auto_config()
    stages = []
    for spec in cfg.stages:  # each is a StageSpec
        stages.append({"use": spec.use, "needs": spec.needs, "on_error": spec.on_error, "config": spec.config})
    return json.dumps({"pipeline": cfg.pipeline, "stages": stages, "decisions": cfg.decisions})


def skip_if_falsy_json(input_str: str) -> str:
    return json.dumps(not json.loads(input_str))

File: packages/typescript/goldenpipe/src/core/wasm/plannerJson.ts
/**
 * plannerJson.ts — the serialize -> backend -> deserialize side of the five
 * planner seams. Each `*ViaWasm` fn serializes the live TS inputs to
 * goldenpipe-core's JSON shapes, calls the registered WASM backend, and
 * reconstructs the SAME rich TS objects the pure-TS planner returns (rehydrating
 * live `Stage` handles from the registry, WiringError, router notes, etc.).
 *
 * Called ONLY from the guarded public entries (Resolver.resolve, Router.apply,
 * the decision gates, computeAutoConfig, the runner's isFalsy call site) when a
 * backend is registered. The pure-core (`*Pure`) bodies never route here, so
 * there is no recursion.
 *
 * NOTE on the import cycle: resolver.ts imports `resolveViaWasm` from here, and
 * here we import the `WiringError` VALUE from resolver.ts. This is a 2-module
 * ESM value cycle, but every reference is inside a function body (never at
 * module-evaluation time), so ESM live bindings resolve it safely. The type
 * deps (PlannedStage/ExecutionPlan) use `import type` so they carry zero runtime
 * weight.
 *
 * Edge-safe: no `node:` imports.
 */

import { WiringError, AmbiguousProducerError, CycleError, UnknownNeedError } from "../engine/resolver.js";
import type { PlannedStage, ExecutionPlan } from "../engine/resolver.js";
import { makeStageSpec, makeDecision, makePipelineConfig } from "../models.js";
import type { Decision, PipeContext, PipelineConfig, StageSpec } from "../models.js";
import type { StageRegistry } from "../engine/registry.js";
import type { PipeWasmBackend } from "./backend.js";

/** Reconstruct the pure-TS error surface from goldenpipe-core's `err` payload. */
function throwFromErr(err: { kind: string; [k: string]: unknown }): never {
  if (err.kind === "missing_producer") {
    throw new WiringError(
      `Stage '${String(err.stage)}' consumes '${String(err.artifact)}' but no prior stage produces it.`,
      { stage: String(err.stage), artifact: String(err.artifact) },
    );
  }
  if (err.kind === "ambiguous_producer") {
    throw new AmbiguousProducerError(String(err.artifact), (err.producers as string[]) ?? []);
  }
  if (err.kind === "cycle") {
    throw new CycleError((err.stages as string[]) ?? []);
  }
  if (err.kind === "unknown_need") {
    throw new UnknownNeedError(String(err.stage), (err.needs as string[]) ?? []);
  }
  if (err.kind === "unknown_stage") {
    throw new Error(`Stage '${String(err.use)}' not found in registry`);
  }
  // Includes "parse" and any future kind.
  throw new Error(`goldenpipe-wasm error kind '${err.kind}': ${JSON.stringify(err)}`);
}

interface SerStage {
  name: string;
  use: string;
  config?: Record<string, unknown>;
  on_error?: string;
  skip_if?: string;
}

export function resolveViaWasm(
  config: PipelineConfig,
  registry: StageRegistry,
  b: PipeWasmBackend,
): ExecutionPlan {
  const all = registry.listAll(); // Record<name, StageInfo> — real registry keys by name, so key===name.
  const stages = Object.entries(all).map(([name, i]) => ({
    key: name,
    name: i.name,
    produces: i.produces,
    consumes: i.consumes,
  }));
  const out = JSON.parse(b.resolveJson(JSON.stringify({ config, stages }))) as
    | { ok: { stages: SerStage[] } }
    | { err: { kind: string; [k: string]: unknown } };
  if ("err" in out && out.err) throwFromErr(out.err);
  const okStages = (out as { ok: { stages: SerStage[] } }).ok.stages;
  const planned: PlannedStage[] = okStages.map((s) => ({
    name: s.name,
    stage: registry.get(s.use), // re-correlate to the live stage
    spec: makeStageSpec({
      use: s.use,
      name: s.name,
      onError: (s.on_error ?? "continue") as StageSpec["onError"],
      ...(s.skip_if !== undefined ? { skipIf: s.skip_if } : {}),
    }),
    config: s.config ?? {},
  }));
  return { stages: planned };
}

export function applyDecisionViaWasm(
  decision: Decision,
  remaining: PlannedStage[],
  ctx: PipeContext,
  registry: StageRegistry,
  b: PipeWasmBackend,
): PlannedStage[] {
  const remSer: SerStage[] = remaining.map((p) => ({
    name: p.name,
    use: p.spec.use,
    config: p.config,
    on_error: p.spec.onError,
    ...(p.spec.skipIf !== undefined ? { skip_if: p.spec.skipIf } : {}),
  }));
  const out = JSON.parse(
    b.applyDecisionJson(JSON.stringify({ decision, remaining: remSer })),
  ) as { remaining: SerStage[]; router_note?: string };
  const byName = new Map(remaining.map((p) => [p.name, p]));
  const next: PlannedStage[] = out.remaining.map((s) => {
    const orig = byName.get(s.name);
    if (orig) return orig; // kept stage: preserve the ORIGINAL live PlannedStage
    // inserted stage: rehydrate a live handle from the registry
    return {
      name: s.name,
      stage: registry.get(s.use),
      spec: makeStageSpec({
        use: s.use,
        name: s.name,
        onError: (s.on_error ?? "continue") as StageSpec["onError"],
      }),
      config: s.config ?? {},
    };
  });
  if (out.router_note !== undefined) ctx.reasoning["_router"] = out.router_note;
  return next;
}

export function evaluateBuiltinViaWasm(
  name: string,
  ctx: PipeContext,
  b: PipeWasmBackend,
): Decision | null {
  const out = JSON.parse(
    b.evaluateBuiltinJson(
      JSON.stringify({ name, ctx: { artifacts: ctx.artifacts, metadata: ctx.metadata } }),
    ),
  ) as { skip: string[]; abort: boolean; insert: string[]; reason: string } | null;
  if (out === null) return null;
  return makeDecision({ skip: out.skip, abort: out.abort, insert: out.insert, reason: out.reason });
}

export function autoConfigViaWasm(
  registry: StageRegistry,
  identityOpts: Record<string, unknown>,
  b: PipeWasmBackend,
): PipelineConfig {
  const available = Object.keys(registry.listAll());
  const out = JSON.parse(
    b.autoConfigJson(JSON.stringify({ available, identity_opts: identityOpts })),
  ) as {
    pipeline: string;
    stages: Array<{ use: string; needs?: string[]; on_error?: string; config?: Record<string, unknown> }>;
    decisions?: string[];
  };
  const stages: StageSpec[] = out.stages.map((s) =>
    makeStageSpec({
      use: s.use,
      needs: s.needs ?? [],
      onError: (s.on_error ?? "continue") as StageSpec["onError"],
      config: s.config ?? {},
    }),
  );
  return makePipelineConfig({ pipeline: out.pipeline, stages, decisions: out.decisions ?? [] });
}

export function skipIfFalsyViaWasm(value: unknown, b: PipeWasmBackend): boolean {
  return JSON.parse(b.skipIfFalsyJson(JSON.stringify(value))) as boolean;
}
